module/builder/collection/builder.go

```diff
@@ -60,277 +53,277 @@ func NewBuilder(
-			if errors.Is(err, storage.ErrNotFound) {
-				continue // in case we are configured with liberal transaction ingest rules
-			}
-			if err != nil {
-				return fmt.Errorf("could not retrieve reference header: %w", err)
-			}
-
-			// for now, disallow un-finalized reference blocks
-			if refChainFinalizedHeight < refHeader.Height {
-				continue
-			}
-
-			// ensure the reference block is not too old
-			txID := tx.ID()
-			if refChainFinalizedHeight-refHeader.Height > uint64(flow.DefaultTransactionExpiry-b.config.ExpiryBuffer) {
-				// the transaction is expired, it will never be valid
-				b.transactions.Rem(txID)
-				continue
-			}
-
-			// check that the transaction was not already used in un-finalized history
-			if lookup.isUnfinalizedAncestor(txID) {
-				continue
-			}
-
-			// check that the transaction was not already included in finalized history.
-			if lookup.isFinalizedAncestor(txID) {
-				// remove from mempool, conflicts with finalized block will never be valid
-				b.transactions.Rem(txID)
-				continue
-			}
-
-			// enforce rate limiting rules
-			if limiter.shouldRateLimit(tx) {
-				continue
-			}
-
-			// ensure we find the lowest reference block height
-			if refHeader.Height < minRefHeight {
-				minRefHeight = refHeader.Height
-				minRefID = tx.ReferenceBlockID
-			}
-
-			// update per-payer transaction count
-			limiter.transactionIncluded(tx)
-
-			transactions = append(transactions, tx)
-			totalByteSize += txByteSize
-			totalGas += tx.GasLimit
+		// cause the max gas limit per tx is way smaller than the total max gas per collection, we can stop here and not continue.
+		// to make it more effective in the future we can continue adding smaller ones
+		if totalGas+tx.GasLimit > b.config.MaxCollectionTotalGas {
+			break
 		}
 
-		// STEP FOUR: we have a set of transactions that are valid to include
-		// on this fork. Now we need to create the collection that will be
-		// used in the payload and construct the final proposal model
-		// TODO (ramtin): enable this again
-		// b.tracer.FinishSpan(parentID, trace.COLBuildOnCreatePayload)
-		// b.tracer.StartSpan(parentID, trace.COLBuildOnCreateHeader)
-		// defer b.tracer.FinishSpan(parentID, trace.COLBuildOnCreateHeader)
-
-		// build the payload from the transactions
-		payload := cluster.PayloadFromTransactions(minRefID, transactions...)
-
-		header := flow.Header{
-			ChainID:     parent.ChainID,
-			ParentID:    parentID,
-			Height:      parent.Height + 1,
-			PayloadHash: payload.Hash(),
-			Timestamp:   time.Now().UTC(),
-
-			// NOTE: we rely on the HotStuff-provided setter to set the other
-			// fields, which are related to signatures and HotStuff internals
+		// retrieve the main chain header that was used as reference
+		refHeader, err := b.mainHeaders.ByBlockID(tx.ReferenceBlockID)
+		if errors.Is(err, storage.ErrNotFound) {
+			continue // in case we are configured with liberal transaction ingest rules
```

Comment:
Question here:

Once the builder has built a lookup, it will start iterate through `transaction.All()` in the mempool, but is it a problem if there are too many tx to iterate through? the tx mempool has a size cap, so the worst case is the tx mempool being full. 

The reason I think of this is that when a collection node is falling behind mainchain, it will receive lots of txs with unknown `tx.ReferenceBlockID`, and it seems we still keep it in the mempool, this would cause the falling behind collection node to be slow during catch up, when it's entering a view where it needs to propose a block. 

I wonder if make sense to drop a tx if reference block ID is unknown. Do we already check reference block ID before adding it to the mempool, if so, this ErrNotFound case should not happen, otherwise it's a bug.

Reply:
We could two a 2-step approach:
* ideally, all transactions in the mempool should have a _finalized_ block as their reference block (see [here](https://github.com/onflow/flow-go/pull/2385#discussion_r871985113) for reasoning)
* we could cache all TXs whose reference block is _not_ finalized
* when the collector locally confirms that another block on the main chain has been finalized (`OnBlockFinalized` event), it can release the corresponding TXs into the mempool. 
* pruning heurisitic for the cache:

Reply:
The configuration for whether to ingest transactions with unknown reference blocks is in the [`TransactionValidator`](https://github.com/onflow/flow-go/blob/2df4ced13a14d61a6b784490af9b3cf02cc96d07/access/validator.go#L199), instantiated in the [`ingest engine`](https://github.com/onflow/flow-go/blob/57f89d4e96259f08fe84163c91ecd32484401b45/engine/collection/ingest/engine.go#L61). We do not accept transactions with unknown reference blocks in the collection node (`AllowUnknownReferenceBlockID=false`). The builder is tolerant of unknown reference blocks so that it respects the transaction ingest rules set in the `ingest` engine, regardless of what they are.

Reply:
Looks like even though we have this option `AllowUnknownReferenceBlockID`, [the collection node is always using](https://github.com/onflow/flow-go/blob/master/engine/collection/ingest/engine.go#L64) `AllUnknownReferenceBlockID=false`, so it's not an issue.